### PR TITLE
Add step for snapshot deployments

### PR DESCRIPTION
**Pull request by KevinHudemann**
BLI: https://github.com/SAP/cloud-sdk-java-backlog/issues/252
Build with deployment: https://github.com/SAP/cloud-sdk-java/actions/runs/6011750052/job/16305849354

### CODE REVIEW

File: .github/workflows/build.yaml [hunk 1]
```diff
@@ -302,4 +310,61 @@ jobs:
...
+          name: target-dirs
+          path: .
+      - name: UNZIP Target Dirs
+        run: |
+          unzip -D target.zip
+      - name: Publish Snapshot
+        run: >
+          mvn $MVN_CLI_ARGS
+          -DaltSnapshotDeploymentRepository=artifactory-snapshots::https://common.repositories.cloud.sap/artifactory/build-snapshots-cloudsdk
+          -Dmaven.test.skip -Dmaven.compiler.showCompilationChanges -Dhttp.keepAlive=false-Dformatter.skip
```
Comment (Johannes-Schneider) [inline .github/workflows/build.yaml:365]: [typo]

```suggestion
          -Dmaven.test.skip -Dmaven.compiler.showCompilationChanges -Dhttp.keepAlive=false -Dformatter.skip
```

File: .pipeline/scripts/get-maven-excludes.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+def write_maven_excludes_github_output(filter_key, filter_value):
+    with open("module-inventory.json", "r") as file:
+        module_inventory = json.load(file)
+
+        filtered_inventory = filter(lambda x: x[filter_key] == filter_value, module_inventory)
+        artifact_ids = list(map(lambda x: x["artifactId"], filtered_inventory))
+        artifact_ids_for_maven_excludes = list(map(lambda x: "!:" + x, artifact_ids))
+
+        excludes = ','.join(artifact_ids)
+        print("Excludes: ",excludes)
```
Comment (Johannes-Schneider) [inline .pipeline/scripts/get-maven-excludes.py:14]: [minor - code style]

```suggestion
        print("Excludes: ", excludes)
```

File: .github/workflows/build.yaml [hunk 3]
```diff
@@ -302,4 +310,62 @@ jobs:
       - name: Perform CodeQL Analysis
         uses: github/codeql-action/analyze@v2
         with:
-          category: "/language:java"
\ No newline at end of file
+          category: "/language:java"
+
+  deploy-snapshot:
+    name: Deploy Snapshot
+    if: ${{ github.ref == 'refs/heads/main' }}
```
Comment (Johannes-Schneider) [inline .github/workflows/build.yaml:317]: [minor]

This condition seems to be fine for the moment.
However, in the future, there might actually be multiple branches we are deploying snapshots from.
(Right now, we are using the scheme `main/v*` for our major version branches, which all need to be able to deploy snapshots)

File: .pipeline/scripts/get-maven-excludes.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
+import os
+import json
+import argparse
+
+def write_maven_excludes_github_output(filter_key, filter_value):
+    with open("module-inventory.json", "r") as file:
+        module_inventory = json.load(file)
+
+        filtered_inventory = filter(lambda x: x[filter_key] == filter_value, module_inventory)
```
Comment (Johannes-Schneider) [inline .pipeline/scripts/get-maven-excludes.py:9]: This comparison might become a problem when filtering for non-string properties (e.g. boolean values).

```suggestion
        filtered_inventory = filter(lambda x: x[filter_key] == str(filter_value), module_inventory)
```

Comment (KevinHudemann) [reply .pipeline/scripts/get-maven-excludes.py:9]: Good point, but I guess you meant it like this right?
```suggestion
        filtered_inventory = filter(lambda x: str(x[filter_key]) == filter_value, module_inventory)
```
I mean `filter_values` should always be a string as it is a parameter and argparse should parse it as string by default, but the value in the modules-inventory might be a non string property like the `excludeFromBlackDuckScan` property.

Comment (Johannes-Schneider) [reply .pipeline/scripts/get-maven-excludes.py]: Yes, you're right!

**Comment by KevinHudemann** [comment]
> I just had a look at the pipeline logs and it seems like the `mvn deploy` is actually re-compiling the modules:
> 
> Logs
> I'm not sure whether this can be mitigated (sources might simply be flagged as _"stale"_ if their timestamp is "too" old) but maybe using the "unzip and touch" workaround (basically `touch` every file in the unzipped "target" folders so that the last modification timestamp is new enough for maven to not re-compile everything) could be considered.

It is not only recompiling but also rebuilding. In addition, it is actually doing the same in the old pipeline.
The question is I guess what you'd like to have. For the Cloud SDK Pipeline we have built a step (nexusUpload) to avoid rebuilding the artifacts and we could also switch to this step if you'd prefer that.

File: .github/workflows/build.yaml (hunk 3 above)
Comment (KevinHudemann) [reply .github/workflows/build.yaml]: @Johannes-Schneider Sure that's doable. You could use functions like [`contains` or `startsWith`](https://docs.github.com/en/actions/learn-github-actions/expressions#startswith) or simply add an `||` and list the branches for which you want to execute the step.
I'd say this can be changed once there is the need to deploy snapshots for other branches, as well.
What do you think?

**Review by Johannes-Schneider — APPROVED** [review]
LGTM

File: .github/workflows/build.yaml [hunk 5]
```diff
@@ -302,4 +310,62 @@ jobs:
...
+
+      # Target
+      - name: Download Target Dirs
+        uses: actions/download-artifact@v3
+        with:
+          name: target-dirs
+          path: .
+      - name: UNZIP Target Dirs
+        run: |
+          unzip -DD target.zip
```
Comment (Johannes-Schneider) [inline .github/workflows/build.yaml:360]: [minor]

We could revert the change as it doesn't change anything I guess.
That way, the pipeline looks a bit more consistent.

```suggestion
          unzip -D target.zip
```

Comment (KevinHudemann) [reply .github/workflows/build.yaml:360]: It at least avoids the recompile.

File: .github/workflows/build.yaml [hunk 6]
```diff
@@ -302,4 +310,62 @@ jobs:
...
+      - name: Publish Snapshot
+        run: >
+          mvn $MVN_CLI_ARGS
+          -DaltSnapshotDeploymentRepository=artifactory-snapshots::https://common.repositories.cloud.sap/artifactory/build-snapshots-cloudsdk
```
Comment (MatKuhr) [inline .github/workflows/build.yaml:364]: I think you could replace `-Dformatter.skip -Dimpsort.skip` with `-DskipFormatting", this will also skip the maven POM sorter 😉
